applications/sdp/mspi/src/hrt/hrt.c

```diff
@@ -41,6 +55,79 @@ static nrf_vpr_csr_vio_shift_ctrl_t xfer_shift_ctrl = {
 	.in_mode = NRF_VPR_CSR_VIO_MODE_IN_CONTINUOUS,
 };
 
+NRF_STATIC_INLINE bool is_counter_running(uint8_t counter)
+{
+	/* This may not work for higher frequencies but
+	 * there is no other way to check if timer is running.
+	 * Tested it down to counter value 2 which is 21.333333MHz.
+	 */
+	uint16_t cnt = nrf_vpr_csr_vtim_simple_counter_get(counter);
+
+	return cnt != nrf_vpr_csr_vtim_simple_counter_get(counter);
+}
+
+NRF_STATIC_INLINE uint32_t get_next_shift_count(volatile hrt_xfer_t *hrt_xfer_params,
+						uint32_t word_count, uint32_t index)
+{
+	switch (word_count - index) {
+	case 1: /* Last transfer */
+		return SHIFTCNTB_VALUE(hrt_xfer_params->xfer_data[HRT_FE_DATA].last_word_clocks);
+	case 2: /* Last but one transfer */
+		return SHIFTCNTB_VALUE(
+			hrt_xfer_params->xfer_data[HRT_FE_DATA].penultimate_word_clocks);
+	default:
+		return SHIFTCNTB_VALUE(BITS_IN_WORD / hrt_xfer_params->bus_widths.data);
+	}
+}
+
+NRF_STATIC_INLINE void rx_end_procedure_mode_3(volatile hrt_xfer_t *hrt_xfer_params)
+{
+	uint32_t last_word;
+	uint16_t in;
+	uint16_t bus_mask = VIO_SINGLE_BUS_MASK;
+	uint16_t bus_position = D1_VIO;
+
+	/* Read last word that is missing data from 1 clock pulse. */
+	last_word = nrf_vpr_csr_vio_in_buffered_reversed_byte_get();
+
+	nrf_vpr_csr_vio_mode_in_set(NRF_VPR_CSR_VIO_MODE_IN_CONTINUOUS);
+
+	if (hrt_xfer_params->counter_value > LAST_WORD_PROCESSING_DELAY) {
+		nrf_vpr_csr_vtim_count_mode_set(0, NRF_VPR_CSR_VTIM_COUNT_STOP);
+		nrf_vpr_csr_vtim_simple_counter_set(0, hrt_xfer_params->counter_value -
+							       LAST_WORD_PROCESSING_DELAY);
+	}
+	/* Wait for the data to appear in the IN register. */
+	while (nrf_vpr_csr_vio_in_get() == IN_REG_EMPTY) {
+	}
+
+	/* Wait for correct edge bit time. */
+	if (hrt_xfer_params->counter_value > LAST_WORD_PROCESSING_DELAY) {
+		nrf_vpr_csr_vtim_simple_wait_set(0, false, 0);
+	}
+
+	in = nrf_vpr_csr_vio_in_get();
+	nrf_vpr_csr_vio_out_or_set(BIT(CLK_VIO));
+
+	/* Insert data from IN register into last word.
+	 * At this point last_word stores all data from last word apart from last clock cycle.
+	 * "in" variable stores state of all VIOs for the last clock cycle, where:
+	 *  - in SINGLE mode: the data is 1 bit in the position of BIT(D1_VIO)
+	 *  - in QUAD mode: the data is 4 bits starting from BIT(D0_VIO)
+	 */
+	if (hrt_xfer_params->bus_widths.data == 4) {
+		bus_mask = VIO_QUAD_BUS_MASK;
+		bus_position = D0_VIO;
+	}
+
+	last_word = (BSWAP_32(last_word) << hrt_xfer_params->bus_widths.data) |
+		    (in & bus_mask) >> bus_position;
```

Comment:
Nit:
```suggestion
	last_word = (BSWAP_32(last_word) << hrt_xfer_params->bus_widths.data) |
		    ((in & bus_mask) >> bus_position);
```

Reply:
Please apply in https://github.com/nrfconnect/sdk-nrf/pull/21321